authz/audit_logger.go

```diff
@@ -0,0 +1,111 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package authz
+
+import (
+	"context"
+	"sync"
+)
+
+var (
+	mu sync.Mutex
+	// auditLoggerBuilderRegistry is a map from the name to the audit logger builder.
+	// This backs the RegisterAuditLoggerBuilder API.
+	auditLoggerBuilderRegistry = make(map[string]AuditLoggerBuilder)
+)
+
+// RegisterAuditLoggerBuilder registers the builder in a global map
+// using b.Name() as the key.
+// This should only be called during initialization time (i.e. in an init() function).
+// If multiple builders are registered with the same name, the one registered last
+// will take effect.
+func RegisterAuditLoggerBuilder(b AuditLoggerBuilder) {
+	mu.Lock()
+	defer mu.Unlock()
+	auditLoggerBuilderRegistry[b.Name()] = b
+}
+
+// AuditInfo contains information used by the audit logger during an audit logging event.
+type AuditInfo struct {
+	// RPCMethod is the method of the audited RPC, in the format of "/pkg.Service/Method".
+	// For example, "/helloworld.Greeter/SayHello".
+	RPCMethod string `json:"rpc_method,omitempty"`
+	// Principal is the identity of the RPC. Currently it will only be available in
+	// certificate-based TLS authentication.
+	Principal string `json:"principal,omitempty"`
+	// PolicyName is the authorization policy name (or the xDS RBAC filter name).
+	PolicyName string `json:"policy_name,omitempty"`
+	// MatchedRule is the matched rule (or policy name in the xDS RBAC filter). It will be
+	// empty if there is no match.
+	MatchedRule string `json:"matched_rule,omitempty"`
+	// Authorized indicates whether the audited RPC is authorized or not.
+	Authorized bool `json:"bool,omitempty"`
+}
+
+// AuditLoggerConfig defines the configuration for a particular implementation of audit logger.
+type AuditLoggerConfig interface {
+	// Name returns the same name as that returned by its supported builder.
+	Name() string
+	// auditLoggerConfig is a dummy interface requiring users to embed this
+	// interface to implement it.
+	auditLoggerConfig()
+}
+
+// AuditLogger is the interface for an audit logger.
+// An audit logger is a logger instance that can be configured to use via the authorization policy
+// or xDS HTTP RBAC filters. When the authorization decision meets the condition for audit, all the
+// configured audit loggers' Log() method will be invoked to log that event with the AuditInfo.
+// The method will be executed synchronously before the authorization is complete and the call is
+// denied or allowed.
+// Please refer to https://github.com/grpc/proposal/pull/346 for more details about audit logging.
```

Comment:
It's unfortunate this is still pending.  How will we remember to update this when it's merged?  Can we wait until then to merge this?

Reply:
Done.

Reply:
"Done" ^ was for the blank `//`. I added a `TODO` to redirect the link to a merged gRFC. I apologize for accumulating tech debt so early but it'd be nice to unblock some of the implementations (IIUC, we need these interfaces when adding the parsing logic in http rbac filter for the logger config, as well as invoking loggers in rbac engines) before the gRFC is completely finalized (as least [one thing](https://github.com/envoyproxy/envoy/pull/26453) outside gRPC that's blocking it).

But of course we should definitely fix it in this PR for anything still unclear with the APIs themselves.